### DISCUSSION TITLE
How to get installed formulae list without calling `brew list` command?

### DISCUSSION THREAD

**Discussion by xareelee** [Homebrew/discussions#248] [Everyday usage]
I want to write a script to get the full installed formulae list without calling `brew list`. How do I do?

1. Is there any file listing all installed formulae?
    * Or do I need to scan the `Cellar` folder for
    * How to distinguish a formula is linked or unlinked.
2. How do I know where installed formulae are? 
    * Brew doesn't guarantee formulae are installed in `/usr/local/Cellar`.
    * The command `brew --env` doesn't provide `HOMEBREW_CELLAR` path.
3. How do I know what version a formula is using without calling `brew` command?
    * The folder `HOMEBREW_CELLAR/{formula}` may contain more than one version.

**Comment by SMillerDev** [comment]
> Is there any file listing all installed formulae?

There isn't, brew checks the install location for `brew list`
> How to distinguish a formula is linked or unlinked.

You'll need to see if it is symlinked.

> How do I know where installed formulae are?

You could parse it from `brew config`

> How do I know what version a formula is using without calling brew command?
The folder HOMEBREW_CELLAR/{formula} may contain more than one version.

You can check what `HOMEBREW_PREFIX/opt/formula_name` points to in the Cellar.

---
All of this is significantly more work than `brew list` and homebrew internal structures are not guaranteed to be stable.

**Reply by xareelee** [reply]
Thanks for reply.

> You can check what `HOMEBREW_PREFIX/opt/formula_name` points to in the Cellar.

Does brew guarantee all installed formulae could be found in the path `HOMEBREW_PREFIX/opt/formula_name`?

**Reply by SMillerDev** [reply]
> homebrew internal structures are not guaranteed to be stable.

**Reply by xareelee** [reply]
I have another question: how do I find out the revision numbers of installed formulae?

[Docs tells](https://docs.brew.sh/Formula-Cookbook#formulae-revisions) some updates don't bump the version, just increase the revision number.

**Reply by SMillerDev** [reply]
`brew info formula --json | jq .[].revision` will show this.
